### PR TITLE
feat: simplify template management

### PR DESCRIPTION
**Pull request by MatousJobanek**
This PR simplifies the management of the templates that are used for operations over tenant namespaces (provisions/delete/update).
The location of the templates is changed from the external projects ([che](https://github.com/fabric8-services/fabric8-tenant-che), [jenkins](https://github.com/fabric8-services/fabric8-tenant-jenkins), [team](https://github.com/fabric8-services/fabric8-tenant-team)) to the directory `environment/templates/` inside of this `fabric8-tenant` project. The files (templates) located in the directory and their names correspond to namespaces they are used for. All objects (contained in the templates) have been moved from the original ones (downloaded from maven central) just with small cosmetic changes - the result and the number of the objects are same.
The cosmetic changes are:
* some of the annotations that were used by FMP are removed
* the version of the templates corresponds to the latest commit of  `fabric8-tenant` project
* labels were slightly changed to not look like a java project
* the content of `fabric8-kubernetes-user-sa.yml` was removed as it wasn't used anywhere
* there is only one template for both `run` and `stage` namespaces: `fabric8-tenant-deploy.yml` - the name of the deploy environment is set by `DEPLOY_TYPE`
* objects of the templates have been rearranged to reflect the namespaces they are used for and files they are located in

Originally, the in-production testing of the changed templates was done via PRs and by setting a release (Nexus) version in https://openshift.io/_profile/_tenant page. The in-production testing will be still possible just with a small change - it won't be downloaded from Nexus, but directly from the GH PR.  That will be possible by taking `sha` of the latest commit in the PR and setting it on the same page as it is done currently. It will be also possible to change the repo and directory the files are downloaded from so you will not have to create a PR to test a template in production.

For better verification, I captured the requests that are sent to openshift api when the namespaces are being created. I printed out the URL and the body of each request.
* this is content for the original version (from master): https://gist.github.com/MatousJobanek/12d062712410349d1e980369a71f6181
* this is content for the code from this PR: https://gist.github.com/MatousJobanek/3beefc66c8e98b147e68dd9f04188662
you can compare them and see that the only differences are labels, annotations, and version.

TODO:
- [x] change https://openshift.io/_profile/_tenant page to reflect the new parameters - PR is here https://github.com/fabric8-ui/fabric8-ui/pull/3344
- [ ] mark original tenant projects  ([che](https://github.com/fabric8-services/fabric8-tenant-che), [jenkins](https://github.com/fabric8-services/fabric8-tenant-jenkins), [team](https://github.com/fabric8-services/fabric8-tenant-team)) as deplecated

### CODE REVIEW

**Comment by MatousJobanek** [comment]
[test]

**Comment by nurali-techie** [comment]
@MatousJobanek can you please explain (or point to some doc) which explain ..
1. why in first place different project for each template is created.
2. what are the fundamental problems with current approach (separate template project) and how it will be solved with new approach (all template in f8-tenant project).

**Comment by MatousJobanek** [comment]
>why in first place different project for each template is created.

To be honest, I have no idea what was behind this decision.

> what are the fundamental problems with the current approach (separate template project) and how it will be solved with new approach (all template in f8-tenant project).

the most fundamental improvements are:
* get rid of FMP - the f8-tenant-che/jenkins/team projects have Java structure that uses FMP to generate the templates, but the tenant projects don't have anything in common with Java and it brings just a big complexity
* simplification
  * currently if you want to change a version of either che or jenkins you have to push a commit to one repo, make a release, then send PR to tenant service that changes the version and merge it there. Then tenant downloads the templates during build and bundles it into the binary - there are some steps that are redundant and not needed at all
   * the projects have a complex hierarchy with several snippets of the resulting template and it is sometimes hard to follow what the snippets refer to
   * we can easily remove some redundant parts of the templates
   * easier/faster in-prod testing by just referring to PR in f8-tenant project
for more info see a part of Tenant v2 doc that refers to the Jenkins template https://docs.google.com/document/d/1HKC3A1Zwx8PSfKc01s3Ymgjd0qhLtPDbX0wDUHnt1ng/edit#heading=h.evupowqjwhx3

File: configuration/configuration.go [hunk 1]
```diff
@@ -36,15 +36,11 @@ const (
...
-	varOpenshiftTeamVersion            = "openshift.team.version"
-	varOpenshiftTemplateDir            = "openshift.template.dir"
 	varOpenshiftUseCurrentCluster      = "openshift.use.current.cluster"
 	varTemplateJenkinsRootURL          = "template.jenkins.root.url"
-	varTemplateRecommenderExternalName = "template.recommender.external.name"
-	varTemplateRecommenderAPIToken     = "template.recommender.api.token"
-	varTemplateDomain                  = "template.domain"
+	VarTemplateRecommenderExternalName = "template.recommender.external.name"
+	VarTemplateRecommenderAPIToken     = "template.recommender.api.token"
+	VarTemplateDomain                  = "template.domain"
```
Comment (nurali-techie) [inline configuration/configuration.go]: see if you can avoid export of env_var.

File: Makefile [hunk 2]
```diff
@@ -143,14 +143,13 @@ migration/sqlbindata.go: $(GO_BINDATA_BIN) $(wildcard migration/sql-files/*.sql)
 		-nocompress \
 		migration/sql-files
 
-template/bindata.go: $(GO_BINDATA_BIN) $(wildcard template/*.yml)
-	CHE_VERSION=$(CHE_VERSION) JENKINS_VERSION=$(JENKINS_VERSION) TEAM_VERSION=$(TEAM_VERSION) EXPOSCONTROLLER_VERSION=$(EXPOSCONTROLLER_VERSION) go generate template/generate.go
```
Comment (nurali-techie) [inline Makefile]: need to remove all these variables in make file and also corresponding file from project root folder.

File: environment/service.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	}
+	return tmpls
+}
+
+type Service struct {
+	ctx               context.Context
+	templatesRepo     string
+	templatesRepoBlob string
+	templatesRepoDir  string
+}
```
Comment (nurali-techie) [inline environment/service.go]: `ctx` as field, make sense ?? OR ctx should be passed as arg to methods for this service.

File: environment/service.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		templatesRepoBlob: templatesRepoBlob,
+		templatesRepoDir:  templatesRepoDir,
+	}
+}
+
+type EnvData struct {
+	NsType    string
+	Name      string
+	Templates []*Template
+	ExpiresAt *time.Time
```
Comment (nurali-techie) [inline environment/service.go]: seems that `ExpiresAt` not used anywhere.

File: environment/service.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+					return nil, errors.New("missing sub in JWT token")
+				}
+				cheMtParams["IDENTITY_ID"] = id.(string)
+			}
+			cheMtParams["REQUEST_ID"] = log.ExtractRequestID(s.ctx)
+			unixNano := time.Now().UnixNano()
+			cheMtParams["JOB_ID"] = strconv.FormatInt(unixNano/1000000, 10)
+
+			templates = templateNames["che-mt"]
+			templates[0].DefaultParams = cheMtParams
```
Comment (nurali-techie) [inline environment/service.go]: I guess there is no test which execute `if` block. Make sense to add one ??

File: environment/service_test.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+		Reply(200).
+		BodyString(defaultLocationTempl)
+
+	service := environment.NewService(context.Background(), "", "123abc", "")
+
+	// when
+	envData, err := service.GetEnvData("run")
+
+	// then
+	require.NoError(t, err)
```
Comment (nurali-techie) [inline environment/service_test.go]: can we add assert to check template.content string has some part from `defaultLocationTempl`

File: environment/service_test.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	vars := map[string]string{
+		"USER_NAME": "dev",
+		"COMMIT":    "123",
+	}
+	objects, err := envData.Templates[0].Process(vars)
+	require.NoError(t, err)
+	assert.Len(t, objects, 1)
+	assert.Equal(t, environment.GetLabel(objects[0], "test"), "default-location")
+}
```
Comment (nurali-techie) [inline environment/service_test.go]: good that we have test which simulate get_template from git_repo :+1:

File: environment/service.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+		}
+		templates = templateNames[envType]
```
Comment (nurali-techie) [inline environment/service.go]: line 90 should be in `else` part right ?? otherwise it will overwrite what is done in if on line 87.

File: environment/template.go [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+		varProjectAdminUser:      masterUser,
+		varCommit:                commit,
+	}
+
+	return merge(vars, getVariables(config))
+}
+
+// RetrieveUserName returns a safe namespace basename based on a username
+func RetrieveUserName(username string) string {
+	return regexp.MustCompile("[^a-z0-9]").ReplaceAllString(strings.Split(username, "@")[0], "-")
```
Comment (nurali-techie) [inline environment/template.go]: `regexp.MustCompile("[^a-z0-9]")` need to run only once so can we move as var outside func.

File: Makefile [hunk 10]
```diff
@@ -143,14 +143,13 @@ migration/sqlbindata.go: $(GO_BINDATA_BIN) $(wildcard migration/sql-files/*.sql)
...
+environment/generated/templates.go: $(GO_BINDATA_BIN) $(wildcard environment/templates/*.yml)
 	$(GO_BINDATA_BIN) \
-		-o template/bindata.go \
-		-pkg template \
-		-prefix '' \
+		-o environment/generated/templates.go \
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply Makefile]: weird, I'm 100% sure that it put it there... probably wrong rebase. Anyway, good catch, thanks :+1:

[the remark this answers is not in the document]
Comment (MatousJobanek) [reply Makefile]: done in bc424de

File: Makefile (hunk 2 above)
Comment (MatousJobanek) [reply Makefile]: same as the previous one bc424de

File: environment/template.go [hunk 11]
```diff
@@ -0,0 +1,314 @@
+package environment
+
```
Comment (nurali-techie) [inline environment/template.go]: Not sure what's the need of doing all these re-factoring ??
I can see that part of `apply.go` file func are in this `template.go` file.

my point is if you want - rename_file / change_location .. but moving part of logic here to there is adding more efforts without any real benefit.

File: environment/service.go (hunk 3 above)
Comment (MatousJobanek) [reply environment/service.go]: good point :+1:

File: environment/service.go (hunk 5 above)
Comment (MatousJobanek) [reply environment/service.go]: This part of the code was just copy-paste from the original location and there was no test before. I was thinking about the same - that there should be a test case covering it. I was trying to add one, but I failed with setting up the toggles for the test case. If you know how to do it - how to set up the feature in toggles - please tell me and I'll add a test for it...

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (nurali-techie) [reply Makefile]: plz also delete file in project root directory. there are files like - JENKINS_VERSION, EXPOSCONTROLLER_VERSION

File: environment/service_test.go (hunk 6 above)
Comment (MatousJobanek) [reply environment/service_test.go]: I think that this should be enough: https://github.com/fabric8-services/fabric8-tenant/pull/634/files/ba57cf813296621b2405d470509c7f97833ee3e9#diff-1f307f420e41ce4eee63dc682f3640f4R89

File: environment/service.go (hunk 5 above)
Comment (nurali-techie) [reply environment/service.go]: hmm, atm, I have no idea.

File: environment/service.go (hunk 8 above)
Comment (MatousJobanek) [reply environment/service.go]: good catch :+1: thanks 
fixed in cc62221

File: Makefile [hunk 12]
```diff
@@ -193,7 +189,7 @@ clean-generated:
 	-rm -rf ./client
 	-rm -rf ./swagger/
 	-rm -f ./migration/sqlbindata.go
-	-rm -f ./template/bindata.go
+	-rm -f ./environment/generated/templates.go
```
Comment (nurali-techie) [inline Makefile]: deleting file would be enough or deleting folder `generated` would be better.

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply Makefile]: the EXPOSCONTROLLER_VERSION was omitted, but the rest should have been already removed - at least I don't see in locally. Or is it again some wrong rebase?

File: environment/template.go (hunk 11 above)
Comment (MatousJobanek) [reply environment/template.go]: which part are you talking about? I mean which part that would be related to the "apply" operation and not to processing templates or retrieving information from them. This file should contain only the logic that relates to templates and not to apply operation, or am I missing something?

File: Makefile (hunk 12 above)
Comment (MatousJobanek) [reply Makefile]: I kept it consistent - the same thing is done for the other files.  the environment/generated/ dir is in .gitignore anyway

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (nurali-techie) [reply Makefile]: now it's fine :+1:

File: configuration/configuration.go (hunk 1 above)
Comment (MatousJobanek) [reply configuration/configuration.go]: unexported in d574aac

File: environment/template.go (hunk 11 above)
Comment (nurali-techie) [reply environment/template.go]: @MatousJobanek I have ran init_teant_test.go for both before and after this PR and compare response body for each request made during test, similar to what you have captured here ([before_link](https://gist.github.com/MatousJobanek/12d062712410349d1e980369a71f6181), [after_link](https://gist.github.com/MatousJobanek/3beefc66c8e98b147e68dd9f04188662)).

After capturing response_body on my local machine, I compared them and it looks good.  It has few differences wrt annotations and labels which you already mentioned.

So with this exercise, I see the changes are fine :+1:

File: environment/template.go [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+		return objects, err
+	}
+	if paramsFromTemplate != nil {
+		templateVars = merge(paramsFromTemplate, templateVars)
+	}
+	pt, err := t.ReplaceVars(templateVars)
+	if err != nil {
+		return objects, err
+	}
+	t.Version = vars[varCommit]
```
Comment (nurali-techie) [inline environment/template.go]: As discussed in sync meeting, it would be better if we have latest CommitSHA for given template_file rather latest CommitSHA of tenant_service.  Also, check when you getting CommitSHA corresponding to template_file, you can get commit_timestamp also.  If so then rather using CommitSHA, we can consider using commit_timestamp which will be more human_readable as well comparable (wrt previous and latest versioning).

**Comment by nurali-techie** [comment]
> > > > why in first place different project for each template is created.
> > > > To be honest, I have no idea what was behind this decision.
> 
> Here is the rational behind it :
> 
the mail thread - http://post-office.corp.redhat.com/archives/devtools-saas/2017-December/msg00271.html (need VPN)

@MatousJobanek it's worth to read this mail thread.  I will also do so. 
@chmouel thanks :+1:

**Comment by nurali-techie** [comment]
@MatousJobanek as we know, on each PR submitted on template project, there is a link providing in PR to test template changes on openshift.io.  With this PR, how do we achieve the similar stuff.

**Comment by MatousJobanek** [comment]
@nurali-techie 
> as we know, on each PR submitted on template project, there is a link providing in PR to test template changes on openshift.io. With this PR, how do we achieve the similar stuff.

well, there is already a [PR](https://github.com/fabric8-ui/fabric8-ui/pull/3344) that changes the _tenant page. It will work in a similar way as it does now, this means that you can create a URL with parameters that automatically fill the form (eg.: `_profile/_tenant?templatesRepoBlob=3be1be1`). 
The only thing that is missing is a bot that takes the `sha` of the latest commit and sends a comment with this link to the PR (only to those PRs that change any of the templates of course :-)). For that, we can use either CI job or create a new simple [Ike](https://github.com/arquillian/ike-prow-plugins/) plugin or maybe there is some bot in the Github marketplace ready for use.

**Comment by chmouel** [comment]
> The only thing that is missing is a bot that takes the `sha` of the latest commit and sends a comment with this link to the PR (only to those PRs that change any of the templates of course :-)). For that, we can use either CI job or create a new simple [Ike](https://github.com/arquillian/ike-prow-plugins/) plugin or maybe there is some bot in the Github marketplace ready for use.

Why we just don't do on CI ? Detect what have been changed, if it's a template file then send a comment to the PR with this link (this is how it is done in with the nexus/tempaltes)  : 

https://github.com/fabric8io/openshift-jenkins-s2i-config/blob/32a9765854bb1866104fb475bde5dd8e872b1bc1/.cico/setup.sh#L44-L54

**Comment by MatousJobanek** [comment]
> Why we just don't do on CI 

yes, that would be the easiest solution - that's why I mentioned it as the first one :-)

> ... (this is how it is done in with the nexus/tempaltes) :
https://github.com/fabric8io/openshift-jenkins-s2i-config/blob/32a9765854bb1866104fb475bde5dd8e872b1bc1/.cico/setup.sh#L44-L54

I didn't know about this script. Thanks for the link :+1:

**Comment by chmouel** [comment]
> > Why we just don't do on CI
> 
> yes, that would be the easiest solution - that's why I mentioned it as the first one :-)

Oups sorry i missed this,

> I didn't know about this script. Thanks for the link 👍

Just FYI, You will need to have a github token for this which is setup here : 

https://github.com/openshiftio/openshiftio-cico-jobs/blob/58f964df6df7ad5814a928a0ec1a8439ca2fdad1/devtools-ci-index.yaml#L1501

File: Gopkg.lock [hunk 14]
```diff
@@ -539,6 +539,14 @@
   pruneopts = "UT"
   revision = "90fa682c2a6e6a37b3a1364ce2fe1d5e41af9d6d"
 
+[[projects]]
```
Comment (alexeykazakov) [inline Gopkg.lock]: It looks like you are editing Gopkg.lock manually? You are not supposed to :) See https://golang.github.io/dep/docs/Gopkg.lock.html

Use `dep ensure` instead.

File: utils/url.go [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+	}
+
+	defer func() {
+		if cerr := resp.Body.Close(); cerr != nil && err == nil {
+			err = cerr
+		}
+	}()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply utils/url.go]: Good point :-) 
fixed 1c3e971

File: openshift/config.go [hunk 16]
```diff
@@ -4,52 +4,66 @@ import (
...
+	TemplatesRepoDir  string
 }
 
 // NewConfig builds openshift config for every user request depending on the user profile
-func NewConfig(baseConfig Config, user *authclient.UserDataAttributes, clusterUser, clusterToken, clusterURL string) Config {
-	return overrideTemplateVersions(user, baseConfig.WithMasterUser(clusterUser).WithToken(clusterToken).WithMasterURL(clusterURL))
+func NewConfig(config *configuration.Data, user *authclient.UserDataAttributes, clusterUser, clusterToken, clusterURL, commit string) Config {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: config.APIServerInsecureSkipTLSVerify(),
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply openshift/config.go]: I don't know to be honest :-). I just moved it from this [location](https://github.com/fabric8-services/fabric8-tenant/blob/master/main.go#L141-L145) - it is used for the same calls as it was before.

File: Gopkg.lock (hunk 14 above)
Comment (MatousJobanek) [reply Gopkg.lock]: Not at all. I know that this file shouldn't be edited manually and I have never done this (at least not intentionally). 
This section was added as part of `dep ensure` process - it is because I started using a new project for mocking HTTP calls in tests. It is [gock](https://github.com/h2non/gock) and from my point of view, it is really powerful and easy to use. My idea is to migrate all tests (in the tenant project) that use `go-vcr` to `gock` in the future. For now, I just started using it as part of this PR.

Comment (nurali-techie) [reply Gopkg.lock]: @MatousJobanek I see no dependency added in `Gopkg.toml` file for `gopkg.in/h2non/gock.v1`.  I guess, we required to add it, right?

Comment (MatousJobanek) [reply Gopkg.lock]: I'm not sure if we required, but I definitely wanted to add it there, but I either forgot or it got lost after the several rebase/merge operations. 
Anyway, it's been added here: 38cf717
Thanks for pointing it out

**Comment by MatousJobanek** [comment]
As there were a few updates in the templates (in those versions updated in tenant master branch that I merged into this PR) I made another comparison of both version - before and after PR. I updated the gists that are referenced in the description of this PR with the latest versions. If nothing slipped out of my notice, everything should be good as the results are same.

The [PR](https://github.com/fabric8-ui/fabric8-ui/pull/3344) that updates the `_profile/_tenant` page is approved and ready to merge.

As for the template versioning. We agreed on having `sha` of the latest commit that changes the particular template file as its version. As I don't want to make this PR bigger and bigger, I created another one that implements this changes: https://github.com/fabric8-services/fabric8-tenant/pull/648

File: openshift/config.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply openshift/config.go]: @aslakknutsen ?

File: Gopkg.lock (hunk 14 above)
Comment (alexeykazakov) [reply Gopkg.lock]: BTW gock looks interesting. Might be a better alternative to go-vcr indeed.
